extendr-api/src/iter.rs

```diff
@@ -89,14 +77,21 @@ impl Iterator for StrIter {
             let vector = self.vector.get();
             if i >= self.len {
                 None
-            } else if TYPEOF(vector) == SEXPTYPE::STRSXP {
-                Some(str_from_strsxp(vector, i as isize))
-            } else if TYPEOF(vector) == SEXPTYPE::INTSXP && TYPEOF(self.levels) == SEXPTYPE::STRSXP
-            {
-                let j = *(INTEGER(vector).add(i));
-                Some(str_from_strsxp(self.levels, j as isize - 1))
             } else if TYPEOF(vector) == SEXPTYPE::NILSXP {
-                Some(<&str>::na())
+                None
+            } else if TYPEOF(vector) == SEXPTYPE::STRSXP {
+                str_from_strsxp(vector, i)
+            } else if TYPEOF(vector) == SEXPTYPE::CHARSXP {
+                rstr::charsxp_to_str(vector)
+            } else if Rf_isFactor(vector).into() {
+                // factor support: factor is an integer, and we need
+                // the value of it, to retrieve the assigned label
+                let j = std::slice::from_raw_parts(INTEGER(vector), self.len as _);
```

Comment:
Can we have a meaningful name for this? It is not a trivial loop index.

Reply:
I've renamed it to `level_index`.